src/web/wms/src/main/java/org/geoserver/wms/web/data/AbstractStylePage.java

```diff
@@ -291,6 +443,75 @@ public void onSubmit(AjaxRequestTarget target, Form<?> form) {
+                new CodeMirrorEditor.CustomButtonAction() {
+                    @Override
+                    public void onClick(AjaxRequestTarget target) {
+                        dialog.setTitle(new ParamResourceModel("insertImage", getPage()));
+                        dialog.setInitialWidth(385);
+                        dialog.setInitialHeight(175);
+                        dialog.showOkCancel(target, new GeoServerDialog.DialogDelegate() {
+
+                            private ChooseImagePanel imagePanel;
+
+                            @Override
+                            protected Component getContents(String id) {
+                                return imagePanel = new ChooseImagePanel(id,
+                                        style.getWorkspace());
+                            }
+
+                            @Override
+                            protected boolean onSubmit(AjaxRequestTarget target, Component contents) {
+                                String imageFileName = imagePanel.getChoice();
+                                if (Strings.isEmpty(imageFileName)) {
+                                    FileUpload fu = imagePanel.getFileUpload();
+                                    imageFileName = fu.getClientFileName();
+                                    int teller = 0;
+                                    GeoServerDataDirectory dd =
+                                            GeoServerApplication.get().getBeanOfType(GeoServerDataDirectory.class);
+                                    Resource res = dd.getStyles(style.getWorkspace(), imageFileName);
+                                    while (Resources.exists(res)) {
+                                        imageFileName = FilenameUtils.getBaseName(fu.getClientFileName())
+                                                + "." + (++teller) + "." + 
+                                                FilenameUtils.getExtension(fu.getClientFileName());
+                                        res = dd.getStyles(style.getWorkspace(), imageFileName);
+                                    }
+                                    try(InputStream is = fu.getInputStream()) {
+                                        try(OutputStream os = res.out()) {
+                                            IOUtils.copy(is, os);                                            
+                                        }                                        
+                                    } catch (IOException e) {
+                                        error(e.getMessage());
+                                        target.add(imagePanel.getFeedback());
+                                        return false;
+                                    }
+                                }
+                                
+                                StringBuffer sb = new StringBuffer("<ExternalGraphic>\\n")
```

Comment:
This should be pluggable to support different formats (CSS, YSLD, MBStyle, etc.).

Converting SLD snippents isn't really possible, so it will have to have a custom implementation per format. Something similar to StyleGenerator would probably work, with a bit of extra tweaking.

Reply:
What about just having the code insert the path to the image, instead of the whole XML block around it? That should work well across formats.

Reply:
But that would also reduce the convenience of the functionality. Perhaps a pluggable option with default (not SLD): only path is the best compromise for now?

Reply:
@NielsCharlier That seems like a good option, then module maintainers can implement there own for other formats

Reply:
Agreed